### PR TITLE
Create Address::matches_script_pubkey method

### PR DESCRIPTION
**Pull request by hashmap**
to check if an address creates a particular script without allocating.

fixes rust-bitcoin/rust-bitcoin#1604

### CODE REVIEW

File: bitcoin/src/address.rs [hunk 1]
```diff
@@ -469,6 +469,28 @@ impl Payload {
...
+                false
+            }
+            Payload::WitnessProgram(ref prog) => if script.is_witness_program() {
+                &script.as_bytes()[2..] == prog.program.as_bytes()
+            } else {
+                false
+            },
+        }
+
+    }
```
Comment (tcharding) [inline bitcoin/src/address.rs:492]: Perhaps this is cleaner?

```rust
    pub fn matches_script_pubkey(&self, script: &Script) -> bool {
        match *self {
            Payload::PubkeyHash(ref hash) => if script.is_p2pkh() {
                return &script.as_bytes()[3..23] == <PubkeyHash as AsRef<[u8; 20]>>::as_ref(hash);
            }
            Payload::ScriptHash(ref hash) => if script.is_p2sh() {
                return &script.as_bytes()[2..22] == <ScriptHash as AsRef<[u8; 20]>>::as_ref(hash);
            }
            Payload::WitnessProgram(ref prog) => if script.is_witness_program() {
                return &script.as_bytes()[2..] == prog.program.as_bytes();
            },
        }
        false
    }
```

**Review by tcharding — DISMISSED** [review]
Thanks for the contribution @hashmap! My suggestion is just a minor one.

ACK 683aab3abe42e8c5120dcee43d12cc84af492637

File: bitcoin/src/address.rs (hunk 1 above)
Comment (hashmap) [reply bitcoin/src/address.rs:492]: It's shorter for sure, but I personally not a fan of multiple returns. This fn is small though, so why not.

**Comment by hashmap** [comment]
hm, is it a known issue? looks unrelated to this change
```
error[E0658]: use of unstable library feature 'matches_macro'
    --> /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/syn-1.0.108/src/lit.rs:1121:32
     |
1121 | ...                   if matches!(b, b' ' | b'\t' | b'\n' | b'\r') {
     |                          ^^^^^^^
     |
     = note: for more information, see https://github.com/rust-lang/rust/issues/65721
```

**Comment by tcharding** [comment]
You can rebase now and that matches issue is fixed

File: bitcoin/src/address.rs (hunk 1 above)
Comment (tcharding) [reply bitcoin/src/address.rs:492]: As you wish man, no bother either way.

File: bitcoin/src/address.rs [hunk 2]
```diff
@@ -469,6 +469,28 @@ impl Payload {
         }
     }
 
+    /// Returns true if the address creates a particular script
```
Comment (Kixunil) [inline bitcoin/src/address.rs:472]: A comment about not allocating would be helpful.

File: bitcoin/src/address.rs (hunk 1 above)
Comment (Kixunil) [reply bitcoin/src/address.rs:492]: Or this:

```rust
    pub fn matches_script_pubkey(&self, script: &Script) -> bool {
        match *self {
            Payload::PubkeyHash(ref hash) if script.is_p2pkh() => {
                &script.as_bytes()[3..23] == <PubkeyHash as AsRef<[u8; 20]>>::as_ref(hash)
            },
            Payload::ScriptHash(ref hash) if script.is_p2sh() => {
                &script.as_bytes()[2..22] == <ScriptHash as AsRef<[u8; 20]>>::as_ref(hash)
            },
            Payload::WitnessProgram(ref prog) if script.is_witness_program() => {
                &script.as_bytes()[2..] == prog.program.as_bytes()
            },
            Payload::PubkeyHash(_) | Payload::ScriptHash(_) | Payload::WitnessProgram(_) => false,
        }
    }
```

File: bitcoin/src/address.rs [hunk 3]
```diff
@@ -955,6 +977,11 @@ impl Address {
         let payload = self.payload.inner_prog_as_bytes();
         payload == xonly_pubkey.serialize()
     }
+
+    /// Returns true if the address creates a particular script
```
Comment (Kixunil) [inline bitcoin/src/address.rs:981]: A comment about not allocating would be helpful.

**Comment by tcharding** [comment]
Ha! This round goes to Kix :)

File: bitcoin/src/address.rs (hunk 2 above)
Comment (hashmap) [reply bitcoin/src/address.rs:472]: thanks, added

File: bitcoin/src/address.rs (hunk 3 above)
Comment (hashmap) [reply bitcoin/src/address.rs:976]: added

File: bitcoin/src/address.rs (hunk 1 above)
Comment (hashmap) [reply bitcoin/src/address.rs]: thank you all for suggestions! updated

**Review by Kixunil — APPROVED** [review]
ACK d71c31c2357b09ecc0daad1a61e0117fd075576a

**Review by apoelstra — APPROVED** [review]
ACK d71c31c2357b09ecc0daad1a61e0117fd075576a

**Pull request merged**